src/sage/schemes/elliptic_curves/kummer_line.py

```diff
@@ -0,0 +1,1765 @@
+        DA = D * A
+        CB = C * B
+        t0 = DA + CB
+        t1 = t0 * t0
+        XQP = ZPQ * t1
+        t2 = DA - CB
+        t3 = t2 * t2
+        ZQP = XPQ * t3
+        X2P = AA * BB
+        t4 = d * E
+        t5 = BB + t4
+        Z2P = E * t5
+
+        return X2P, Z2P, XQP, ZQP
+
+
+class KummerLinePoint(SageObject):
+    r"""
+    Kummer point on a Kummer line.
+
+    EXAMPLES::
+
+        sage: E = EllipticCurve(GF(101), [2, 3])
+        sage: K = KummerLine(E)
+        sage: P = E(95, 52)
+
+    A Kummer point can be constructed from `XZ`-coordinates::
+
+        sage: xP = KummerLinePoint(K, [95, 1]); xP
+        (95 : 1)
+
+    Or, it can be constructed from a point on the elliptic curve::
+
+        sage: xP = KummerLinePoint(K, P); xP
+        (95 : 1)
+
+    The point can also be created by calling the Kummer line::
+
+        sage: xP = K(P); xP
+        (95 : 1)
+    """
+
+    def __init__(self, parent, coords):
+        r"""
+        Create a point from the coordinates.
+
+        INPUT:
+
+            - ``parent`` - A Kummer line
+
+            - ``coords`` - either a point P on EllipticCurve or a tuple (X, Z) where `P = (X : * : Z)`; Z is optional
+
+        EXAMPLES::
+
+            sage: E = EllipticCurve(GF(101), [2, 3])
+            sage: P = E(95, 52)
+            sage: K = KummerLine(E)
+
+        A Kummer point can be constructed from `XZ`-coordinates::
+
+            sage: xP = KummerLinePoint(K, [95, 1]); xP
+            (95 : 1)
+
+        Z-coordinate is optional, assumed to be 1 then::
+
+            sage: KummerLinePoint(K, 95) == KummerLinePoint(K, [95, 1])
+            True
+            sage: KummerLinePoint(K, 95) == KummerLinePoint(K, [95, 2])
+            False
+            sage: KummerLinePoint(K, 95) == KummerLinePoint(K, [95])
+            True
+
+        Or, it can be constructed from a point on the elliptic curve::
+
+            sage: xP = KummerLinePoint(K, P); xP
+            (95 : 1)
+        """
+        # Ensure the parent is the right type
+        if not isinstance(parent, KummerLine_generic):
+            raise TypeError("not a Kummer line")
+
+        R = parent.base_ring()
+
+        # Point at infinity
+        if coords is None:
+            coords = (R(1), R(0))
+
+        # Construct point from P on an elliptic curve in Montgomery form
+        elif isinstance(coords, EllipticCurvePoint_field):
+            # Make sure point's parent curve matches with Kummer Line
+            if coords.is_zero():
+                coords = (R(1), R(0))
+            else:
+                a, b = parent.extract_weierstrass_constants()
+                # assert coords.curve().a_invariants() == (0, 0, 0, a, b)
+                # Done this way to handle Montgomery curves
+                assert coords.curve().is_isomorphic(EllipticCurve(R, [a, b]))
+                coords = coords[0], coords[2]
+
+        # Construct from X coordinate only
+        elif isinstance(coords, RingElement):
```

Comment:
I think this is better handled if you do a `try` cast into `R`, otherwise say `K(int(1))` will fail. I know this is stupid but I write code in `.py` a lot :(